[WebGPU] Introduce WebGPUSurface and implement GPUCanvasContext on top of it

tuankiet65 opened the pull request:
#### 53198d81a9e7bc8762b40d7782cb24260a1f8ba6
<pre>
[WebGPU] Introduce WebGPUSurface and implement GPUCanvasContext on top of it
<a href="https://bugs.webkit.org/show_bug.cgi?id=241881">https://bugs.webkit.org/show_bug.cgi?id=241881</a>
&lt;rdar://96172732&gt;

Reviewed by NOBODY (OOPS!).

This commit introduces WebGPUSurface, representing a WebGPU surface / swapchain that
can be drawn into by JavaScript applications using WebGPU. GPUCanvasContext is then
implemented by delegating to WebGPUSurface.

WebGPUSurface represents a generic WebGPU surface / swapchain. The available interfaces allows
for getting the current WebGPU texture to draw into, and presenting the current texture. The
intention is for different platforms to have their own WebGPUSurface implementation. One
concrete implementation is provided: WebGPUSurfaceCocoa for Apple platforms. Because
WebGPUSurfaceCocoa uses IOSurfaces under the hood, WebGPU.framework and PAL is extended to allow
creating IOSurface-backed textures.

GPUCanvasContext is then implemented as a thin wrapper above WebGPUSurface. When a GPUCanvasContext is
configured or resized, a new WebGPUSurface is created. When a JavaScript application requests the current
texture from GPUCanvasContext, GPUCanvasContext delegates this to the underlying WebGPUSurface.
When a GPUCanvasContext is unconfigured/tear down, the WebGPUSurface is destroyed.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/WebGPU/GPU.cpp:
(WebCore::GPU::getPreferredCanvasFormat):
* Source/WebCore/Modules/WebGPU/GPUCanvasContext.cpp:
(WebCore::getCanvasSizeAsIntSize):
(WebCore::GPUCanvasContext::create):
(WebCore::GPUCanvasContext::GPUCanvasContext):
(WebCore::GPUCanvasContext::reshape):
(WebCore::GPUCanvasContext::canvas):
(WebCore::GPUCanvasContext::configure):
(WebCore::GPUCanvasContext::getCurrentTexture):
(WebCore::GPUCanvasContext::unconfigure):
(WebCore::GPUCanvasContext::pixelFormat const):
(WebCore::GPUCanvasContext::colorSpace const):
(WebCore::GPUCanvasContext::layerContentsDisplayDelegate):
(WebCore::GPUCanvasContext::prepareForDisplay):
(WebCore::GPUCanvasContext::markContextChangedAndNotifyCanvasObservers):
* Source/WebCore/Modules/WebGPU/GPUCanvasContext.h:
(WebCore::GPUCanvasContext::create): Deleted.
* Source/WebCore/Modules/WebGPU/GPUCanvasContext.idl:
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUDeviceImpl.cpp:
(PAL::WebGPU::DeviceImpl::createIOSurfaceBackedTexture):
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUDeviceImpl.h:
* Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUDevice.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::getCSSCanvasContext):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Document.idl:
* Source/WebCore/html/HTMLCanvasElement.cpp:
(WebCore::HTMLCanvasElement::getContext):
(WebCore::HTMLCanvasElement::isWebGPUType):
(WebCore::HTMLCanvasElement::createContextWebGPU):
(WebCore::HTMLCanvasElement::getContextWebGPU):
* Source/WebCore/html/HTMLCanvasElement.h:
* Source/WebCore/html/HTMLCanvasElement.idl:
* Source/WebCore/html/canvas/CanvasRenderingContext.h:
(WebCore::CanvasRenderingContext::isWebGPU const):
* Source/WebCore/platform/graphics/gpu/WebGPUSurface.h: Copied from Source/WebCore/Modules/WebGPU/GPUCanvasContext.cpp.
(WebCore::WebGPUSurface::preferredTextureFormat):
* Source/WebCore/platform/graphics/gpu/cocoa/WebGPUSurfaceCocoa.cpp: Added.
(WebCore::WebGPUSurfaceCocoa::WebGPUSurfaceCocoa):
(WebCore::WebGPUSurfaceCocoa::prepareToDelegateDisplay):
(WebCore::WebGPUSurfaceCocoa::display):
(WebCore::WebGPUSurfaceCocoa::orientation const):
(WebCore::WebGPUSurfaceCocoa::present):
(WebCore::WebGPUSurfaceCocoa::isValidBuffer):
(WebCore::WebGPUSurfaceCocoa::ensureValidBackBuffer):
(WebCore::WebGPUSurfaceCocoa::createNewSurface):
(WebCore::WebGPUSurfaceCocoa::currentTexture):
(WebCore::WebGPUSurfaceCocoa::pixelFormat const):
(WebCore::WebGPUSurfaceCocoa::colorSpace const):
* Source/WebCore/platform/graphics/gpu/cocoa/WebGPUSurfaceCocoa.h: Added.
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::shouldSetContentsDisplayDelegate const):
(WebCore::RenderLayerBacking::updateConfiguration):
* Source/WebCore/rendering/RenderLayerBacking.h:
* Source/WebGPU/WebGPU/Device.h:
* Source/WebGPU/WebGPU/Texture.mm:
(WebGPU::Device::createTexture):
(WebGPU::pixelFormat): Deleted.
(WebGPU::depthOnlyAspectMetalFormat): Deleted.
(WebGPU::stencilOnlyAspectMetalFormat): Deleted.
* Source/WebGPU/WebGPU/WebGPUExt.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.cpp:
(WebKit::RemoteDevice::createIOSurfaceBackedTexture):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.messages.in:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteDeviceProxy.cpp:
(WebKit::WebGPU::RemoteDeviceProxy::createIOSurfaceBackedTexture):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteDeviceProxy.h:
</pre>

youssefsoli commented on Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUDeviceImpl.cpp:
```diff
@@ -130,6 +130,49 @@ Ref<Texture> DeviceImpl::createTexture(const TextureDescriptor& descriptor)
+        label.data(),
+        m_convertToBackingContext->convertTextureUsageFlagsToBacking(descriptor.usage),
+        m_convertToBackingContext->convertToBacking(descriptor.dimension),
+        size,
```
`WTFMove(size)`

tuankiet65 replied:
Fixed.

youssefsoli commented on Source/WebCore/html/HTMLCanvasElement.cpp:
```diff
@@ -538,6 +555,49 @@ ImageBitmapRenderingContext* HTMLCanvasElement::getContextBitmapRenderer(const S
+
+    m_context = GPUCanvasContext::create(*this);
+
+    if (m_context) {
+        // Adds the current document as an observer, so that the document calls prepareForDisplay
+        // when the canvas changes.
+        addObserver(document());
+
+        // Need to make sure a RenderLayer and compositing layer get created for the Canvas.
+        invalidateStyleAndLayerComposition();
+    }
+
+    return static_cast<GPUCanvasContext*>(m_context.get());
```
Do you need the null check?

tuankiet65 replied:
Probably, we don't want to do unnecessary work if we can't create a context. WebGL does similarly.

youssefsoli replied:
Ok but `GPUCanvasContext::create` never returns null unlike the WebGL sites.
```
std::unique_ptr<GPUCanvasContext> GPUCanvasContext::create(CanvasBase& canvas)
 {
     return std::unique_ptr<GPUCanvasContext>(new GPUCanvasContext(canvas));
 }
 ```

tuankiet65 replied:
Hmm okay, it's possible to create a non-functional `GPUCanvasContext` because it's running on platforms that don't support creating `WebGPUSurface` yet. I'll see if we can fix it so that it will refuse to create a GPUCanvasContext on those platforms.

youssefsoli commented on Source/WebCore/html/HTMLCanvasElement.cpp:
```diff
@@ -538,6 +555,49 @@ ImageBitmapRenderingContext* HTMLCanvasElement::getContextBitmapRenderer(const S
+    if (!document().settings().webGPU())
+        return nullptr;
+
+    if (m_context && !m_context->isWebGPU())
+        return nullptr;
```
Maybe an `ASSERT_NOT_REACHED` here since it seems like a very unlikely scenario that should probably be hit in debug mode.

tuankiet65 replied:
If I interpret this correctly, one `<canvas>` should only ever have one context. If it already has a WebGL context, then trying to get a WebGPU context out of it should fail gracefully (returning null on JavaScript side)

tuankiet65 replied:
https://developer.mozilla.org/en-US/docs/Web/API/HTMLCanvasElement/getContext:

> The HTMLCanvasElement.getContext() method returns a drawing context on the canvas, or null if the context identifier is not supported, or the canvas has already been set to a different context mode.

> Later calls to this method on the same canvas element, with the same contextType argument, will always return the same drawing context instance as was returned the first time the method was invoked. It is not possible to get a different drawing context object on a given canvas element.

youssefsoli replied:
But isn't that context already checked at the call site? If checking it again yields a different result then there is a major issue.

tuankiet65 replied:
Only place where `::getContextWebGPU` (where that code belongs) is used is `::getContext`, `::getContext` only checks the `type` string and redirects it to the corresponding method. Or maybe I miss somewhere else :/

youssefsoli replied:
Yup, only the type is checked at those sites you're right.

youssefsoli commented on Source/WebCore/Modules/WebGPU/GPUCanvasContext.cpp:
```diff
@@ -28,26 +28,144 @@
 }
 
-void GPUCanvasContext::configure(const GPUCanvasConfiguration&)
+void GPUCanvasContext::configure(GPUCanvasConfiguration&& configuration)
```
You can make this `GPUCanvasContext::configure(GPUCanvasConfiguration configuration)` to support both a move and a copy if needed by implementors.

tuankiet65 replied:
Fixed.

youssefsoli commented on Source/WebCore/platform/graphics/gpu/cocoa/WebGPUSurfaceCocoa.cpp:
```diff
@@ -0,0 +1,149 @@
+    if (!buffer)
+        return false;
+
+    if (buffer->size() != IntSize { m_width, m_height })
```
Maybe you can store both the width and height in member `IntSize` to avoid creating a new `IntSize` every time this inexpensive function is called.

tuankiet65 replied:
I have a raging belief that this should compiles down to two int comparisons. https://godbolt.org/z/Gofqb7j9j.

cdumez replied:
Yes, I don't think there is a perf difference here. It's just ugly to have to construct an IntSize here. The code would look better here if we stored the data member as an IntSize and thus I think we should probably do it unless there is a good reason not to?

tuankiet65 replied:
That sounds good because we have to create `IntSize`s at two different points in code. Fixed.

youssefsoli commented on Source/WebCore/platform/graphics/gpu/cocoa/WebGPUSurfaceCocoa.cpp:
```diff
@@ -0,0 +1,149 @@
+        return;
+
+    // If the back buffer is not valid, try to recycle the secondary back buffer.
+    m_backBuffer = WTFMove(m_secondaryBackBuffer);
```
Should `m_secondaryBackBuffer` be null checked?

tuankiet65 replied:
It's fine if `m_secondaryBackBuffer` is null, because we check if it's valid in the next line (which checks if it's null first thing)

cdumez commented on Source/WebCore/Modules/WebGPU/GPU.cpp:
```diff
@@ -65,7 +70,11 @@ void GPU::requestAdapter(const std::optional<GPURequestAdapterOptions>& options,
+#if PLATFORM(COCOA)
+    return WebGPUSurfaceCocoa::preferredTextureFormat();
+#else
+    return WebGPUSUrface::preferredTextureFormat();
```
Capitalization issue here

tuankiet65 replied:
Fixed.

cdumez commented on Source/WebCore/Modules/WebGPU/GPUCanvasContext.cpp:
```diff
@@ -28,26 +28,144 @@
 #include "GPUAdapter.h"
 #include "GPUCanvasConfiguration.h"
+#include "RenderBox.h"
+#if PLATFORM(COCOA)
```
Ifdef’d includes come after the other includes, with a blank line in between.

tuankiet65 replied:
Fixed.

cdumez commented on Source/WebCore/Modules/WebGPU/GPUCanvasContext.cpp:
```diff
@@ -28,26 +28,144 @@
+
+GPUCanvasContext::GPUCanvasContext(CanvasBase& canvas)
+    : GPUBasedCanvasRenderingContext(canvas)
+    , m_currentConfiguration()
```
Not needed

tuankiet65 replied:
Fixed.

cdumez commented on Source/WebCore/Modules/WebGPU/GPUCanvasContext.cpp:
```diff
@@ -28,26 +28,144 @@
+GPUCanvasContext::GPUCanvasContext(CanvasBase& canvas)
+    : GPUBasedCanvasRenderingContext(canvas)
+    , m_currentConfiguration()
+    , m_surface()
```
Not needed

tuankiet65 replied:
Fixed.

youssefsoli commented on Source/WebCore/rendering/RenderLayerBacking.cpp:
```diff
@@ -1095,15 +1110,13 @@ bool RenderLayerBacking::updateConfiguration(const RenderLayer* compositingAnces
-#if ENABLE(WEBGL) || ENABLE(OFFSCREEN_CANVAS)
-    else if (renderer().isCanvas() && canvasCompositingStrategy(renderer()) == CanvasAsLayerContents) {
+    else if (shouldSetContentsDisplayDelegate()) {
         const HTMLCanvasElement* canvas = downcast<HTMLCanvasElement>(renderer().element());
```
`auto`

tuankiet65 replied:
I'm kinda hesitant to fix up existing code that works unless necessary

tuankiet65 replied:
Guess I was wrong, the code surrounding it was using `auto` so this kinda sticks out. Changing to `auto*` then.

cdumez commented on Source/WebCore/Modules/WebGPU/GPUCanvasContext.cpp:
```diff
@@ -28,26 +28,144 @@
+    , m_currentConfiguration()
+    , m_surface()
+{
+    suspendIfNeeded();
```
This is wrong, this should be called in the create() function, not in the constructor.

tuankiet65 replied:
Fixed.

youssefsoli commented on Source/WebGPU/WebGPU/Texture.mm:
```diff
@@ -2031,17 +2031,33 @@ static MTLStorageMode storageMode(bool deviceHasUnifiedMemory, bool supportsNonP
+    IOSurfaceRef ioSurfaceBacking = nullptr;
+    const auto* current = descriptor.nextInChain;
+    while (current) {
+        bool viewFormatsSpecified = false;
+        bool ioSurfaceBackingSpecified = false;
```
`auto`

cdumez replied:
How is auto better than bool? Same length, less information.

youssefsoli replied:
Oh interesting, I checked https://webkit.org/code-style-guidelines/ and indeed `auto` isn't a strict rule 👍

cdumez replied:
Right, we do prefer `auto` in general but for longer types. For things like `int` or `bool`, we don't usually use `auto` AFAIK.

youssefsoli commented on Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.cpp:
```diff
@@ -129,6 +129,22 @@ void RemoteDevice::createTexture(const WebGPU::TextureDescriptor& descriptor, We
+    auto surface = WebCore::IOSurface::createFromSendRight(WTFMove(surfaceSendRight), WebCore::DestinationColorSpace::SRGB());
+
+    auto texture = m_backing->createIOSurfaceBackedTexture(*convertedDescriptor, surface->surface());
+    auto remoteTexture = RemoteTexture::create(texture, m_objectHeap, m_streamConnection.copyRef(), identifier);
+    m_objectHeap.addObject(identifier, remoteTexture);
```
`WTFMove` for `texture`, `remoteTexture`, and the last `identifier`

cdumez commented on Source/WebCore/Modules/WebGPU/GPUCanvasContext.cpp:
```diff
@@ -28,26 +28,144 @@
+
+RefPtr<GraphicsLayerContentsDisplayDelegate> GPUCanvasContext::layerContentsDisplayDelegate()
+{
+    if (!m_surface)
```
This if check seems unnecessary. Why not always return m_surface?

tuankiet65 replied:
Fixed.

youssefsoli commented on Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteDeviceProxy.cpp:
```diff
@@ -100,6 +102,25 @@ Ref<PAL::WebGPU::Texture> RemoteDeviceProxy::createTexture(const PAL::WebGPU::Te
+    auto sendResult = send(Messages::RemoteDevice::CreateIOSurfaceBackedTexture(*convertedDescriptor, surfaceSendRight, identifier));
+    UNUSED_VARIABLE(sendResult);
+
+    return RemoteTextureProxy::create(*this, m_convertToBackingContext, identifier);
```
`WTFMove` `surfaceSendRight` and the last `identifier`

cdumez commented on Source/WebCore/Modules/WebGPU/GPUCanvasContext.cpp:
```diff
@@ -28,26 +28,144 @@
+    bool canvasIsDirty = false;
+
+    if (auto* canvas = htmlCanvas()) {
+        RenderBox* renderBox = canvas->renderBox();
```
Auto*

tuankiet65 replied:
Fixed.

cdumez commented on Source/WebCore/Modules/WebGPU/GPUCanvasContext.cpp:
```diff
@@ -28,26 +28,144 @@
+            canvas->clearCopiedImage();
+            // renderBox->contentChanged(CanvasPixelsChanged);
+            renderBox->contentChanged(CanvasChanged);
+            // return;
```
Don’t leave commented code

tuankiet65 replied:
I'm unsure if this is correct so for the time being I'll leave it there and wait for someone to chime in.

youssefsoli commented on Source/WebCore/Modules/WebGPU/GPUCanvasContext.cpp:
```diff
@@ -28,26 +28,144 @@
+{
+    m_compositingResultsNeedsUpdating = true;
+
+    bool canvasIsDirty = false;
```
`auto`

tuankiet65 replied:
There was a discussion above that we should keep this `bool` so I'll keep this.

cdumez commented on Source/WebCore/Modules/WebGPU/GPUCanvasContext.h:
```diff
@@ -25,41 +25,71 @@
-    {
-        return adoptRef(*new GPUCanvasContext());
-    }
+    static std::unique_ptr<GPUCanvasContext> create(CanvasBase&);
```
This seems wrong, the base class (CanvasRenderingContext) is refcounted, so this should return a RefPtr, not a unique ptr

tuankiet65 replied:
All other canvas context creation methods return `std::unique_ptr`, hmm

cdumez replied:
Do they subclass CanvasRenderingContext? CanvasRenderingContext has a ref() / deref() so I don't get it.

cdumez replied:
Never mind, I looked more deeply at the code and using unique_ptr here is likely fine. CanvasRenderingContext's ref() / deref() just ref /deref its owning canvas element. Therefore, the owning canvas need to own the context via a unique_ptr. This is a special pattern.

tuankiet65 replied:
Right, that's what I was thinking

cdumez commented on Source/WebCore/Modules/WebGPU/GPUCanvasContext.h:
```diff
@@ -25,41 +25,71 @@
     void unconfigure();
 
-    RefPtr<GPUTexture> getCurrentTexture();
+    bool isWebGPU() const override { return true; }
```
Should be private

tuankiet65 replied:
If it's public in the parent then should it be public here too? Kinda no point if we make it private, we can just upcast and call it.

cdumez commented on Source/WebCore/Modules/WebGPU/GPUCanvasContext.h:
```diff
@@ -25,41 +25,71 @@
+    bool isWebGPU() const override { return true; }
+    const char* activeDOMObjectName() const override
+    {
+        return "GPUCanvasElement";
```
Should move to the cpp and make private

cdumez commented on Source/WebCore/Modules/WebGPU/GPUCanvasContext.h:
```diff
@@ -25,41 +25,71 @@
 
 private:
-    GPUCanvasContext() = default;
+    GPUCanvasContext(CanvasBase&);
```
Should be marked as explicit

tuankiet65 replied:
Fixed.

cdumez commented on Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUDevice.h:
```diff
@@ -37,6 +37,9 @@
 #include <wtf/HashSet.h>
 #include <wtf/Ref.h>
 #include <wtf/text/WTFString.h>
+#if HAVE(IOSURFACE)
```
Missing blank line before this.

cdumez commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -6586,6 +6587,9 @@ std::optional<RenderingContext> Document::getCSSCanvasContext(const String& type
     if (is<ImageBitmapRenderingContext>(*context))
         return RenderingContext { RefPtr<ImageBitmapRenderingContext> { &downcast<ImageBitmapRenderingContext>(*context) } };
 
+    if (is<GPUCanvasContext>(*context))
```
Good use case for dynamicDowncast

tuankiet65 replied:
The code around the area follows the same `is<...>(...)` pattern, so it'd stand out if we use `dynamicDowncast` here.

cdumez replied:
Yes, it stands out as beautiful :)

cdumez commented on Source/WebCore/platform/graphics/gpu/cocoa/WebGPUSurfaceCocoa.cpp:
```diff
@@ -0,0 +1,149 @@
+    // If the back buffer is not valid, try to recycle the secondary back buffer.
+    m_backBuffer = WTFMove(m_secondaryBackBuffer);
+    m_backBufferTexture = nullptr;
+    m_secondaryBackBuffer = nullptr;
```
You just did a move so this should not be needed.

tuankiet65 replied:
Fixed.

cdumez reviewed (commented):
Incomplete review

cdumez commented on Source/WebCore/platform/graphics/gpu/cocoa/WebGPUSurfaceCocoa.cpp:
```diff
@@ -0,0 +1,149 @@
+    // Can't recycle the secondary back buffer, so just destroy everything and create a new buffer.
+    // a new one.
+    m_backBuffer = nullptr;
+    m_secondaryBackBuffer = nullptr;
```
Shouldn’t this already be null?

tuankiet65 replied:
Fixed.

tuankiet65 committed:
[WebGPU] Introduce WebGPUSurface and implement GPUCanvasContext on top of it
https://bugs.webkit.org/show_bug.cgi?id=241881
<rdar://96172732>

Reviewed by NOBODY (OOPS!).

This commit introduces WebGPUSurface, representing a WebGPU surface / swapchain that
can be drawn into by JavaScript applications using WebGPU. GPUCanvasContext is then
implemented by delegating to WebGPUSurface.

WebGPUSurface represents a generic WebGPU surface / swapchain. The available interfaces allows
for getting the current WebGPU texture to draw into, and presenting the current texture. The
intention is for different platforms to have their own WebGPUSurface implementation. One
concrete implementation is provided: WebGPUSurfaceCocoa for Apple platforms. Because
WebGPUSurfaceCocoa uses IOSurfaces under the hood, WebGPU.framework and PAL is extended to allow
creating IOSurface-backed textures.

GPUCanvasContext is then implemented as a thin wrapper above WebGPUSurface. When a GPUCanvasContext is
configured or resized, a new WebGPUSurface is created. When a JavaScript application requests the current
texture from GPUCanvasContext, GPUCanvasContext delegates this to the underlying WebGPUSurface.
When a GPUCanvasContext is unconfigured/tear down, the WebGPUSurface is destroyed.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/WebGPU/GPU.cpp:
(WebCore::GPU::getPreferredCanvasFormat):
* Source/WebCore/Modules/WebGPU/GPUCanvasContext.cpp:
(WebCore::getCanvasSizeAsIntSize):
(WebCore::GPUCanvasContext::create):
(WebCore::GPUCanvasContext::GPUCanvasContext):
(WebCore::GPUCanvasContext::reshape):
(WebCore::GPUCanvasContext::canvas):
(WebCore::GPUCanvasContext::configure):
(WebCore::GPUCanvasContext::getCurrentTexture):
(WebCore::GPUCanvasContext::unconfigure):
(WebCore::GPUCanvasContext::pixelFormat const):
(WebCore::GPUCanvasContext::colorSpace const):
(WebCore::GPUCanvasContext::layerContentsDisplayDelegate):
(WebCore::GPUCanvasContext::prepareForDisplay):
(WebCore::GPUCanvasContext::markContextChangedAndNotifyCanvasObservers):
* Source/WebCore/Modules/WebGPU/GPUCanvasContext.h:
(WebCore::GPUCanvasContext::create): Deleted.
* Source/WebCore/Modules/WebGPU/GPUCanvasContext.idl:
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUDeviceImpl.cpp:
(PAL::WebGPU::DeviceImpl::createIOSurfaceBackedTexture):
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUDeviceImpl.h:
* Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUDevice.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::getCSSCanvasContext):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Document.idl:
* Source/WebCore/html/HTMLCanvasElement.cpp:
(WebCore::HTMLCanvasElement::getContext):
(WebCore::HTMLCanvasElement::isWebGPUType):
(WebCore::HTMLCanvasElement::createContextWebGPU):
(WebCore::HTMLCanvasElement::getContextWebGPU):
* Source/WebCore/html/HTMLCanvasElement.h:
* Source/WebCore/html/HTMLCanvasElement.idl:
* Source/WebCore/html/canvas/CanvasRenderingContext.h:
(WebCore::CanvasRenderingContext::isWebGPU const):
* Source/WebCore/platform/graphics/gpu/WebGPUSurface.h: Copied from Source/WebCore/Modules/WebGPU/GPUCanvasContext.cpp.
(WebCore::WebGPUSurface::preferredTextureFormat):
* Source/WebCore/platform/graphics/gpu/cocoa/WebGPUSurfaceCocoa.cpp: Added.
(WebCore::WebGPUSurfaceCocoa::WebGPUSurfaceCocoa):
(WebCore::WebGPUSurfaceCocoa::prepareToDelegateDisplay):
(WebCore::WebGPUSurfaceCocoa::display):
(WebCore::WebGPUSurfaceCocoa::orientation const):
(WebCore::WebGPUSurfaceCocoa::present):
(WebCore::WebGPUSurfaceCocoa::isValidBuffer):
(WebCore::WebGPUSurfaceCocoa::ensureValidBackBuffer):
(WebCore::WebGPUSurfaceCocoa::createNewSurface):
(WebCore::WebGPUSurfaceCocoa::currentTexture):
(WebCore::WebGPUSurfaceCocoa::pixelFormat const):
(WebCore::WebGPUSurfaceCocoa::colorSpace const):
* Source/WebCore/platform/graphics/gpu/cocoa/WebGPUSurfaceCocoa.h: Added.
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::shouldSetContentsDisplayDelegate const):
(WebCore::RenderLayerBacking::updateConfiguration):
* Source/WebCore/rendering/RenderLayerBacking.h:
* Source/WebGPU/WebGPU/Device.h:
* Source/WebGPU/WebGPU/Texture.mm:
(WebGPU::Device::createTexture):
(WebGPU::pixelFormat): Deleted.
(WebGPU::depthOnlyAspectMetalFormat): Deleted.
(WebGPU::stencilOnlyAspectMetalFormat): Deleted.
* Source/WebGPU/WebGPU/WebGPUExt.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.cpp:
(WebKit::RemoteDevice::createIOSurfaceBackedTexture):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.messages.in:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteDeviceProxy.cpp:
(WebKit::WebGPU::RemoteDeviceProxy::createIOSurfaceBackedTexture):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteDeviceProxy.h:
